### ISSUE TITLE
support multiple suscription in yaml

### ISSUE THREAD

**Issue by Rocco83** [evgeni/katello-attach-subscription#4]
Hi,

from coding review, seems that we can associate only one subscription in yaml to each entry.
It would be very good to have a set of subscription, associating the first with entitlements available.

From a check, seems that the "activation key approach" cannot be use.
Maybe an array of subscription would work?
What do you think about it?

**Comment by evgeni** [comment]
yeah, I thought about it too, but did not yet have time to implement it properly.

the original need of this script was for hypervisors, those usually only have one sub, but it would be nice to make it more general

**Comment by Rocco83** [comment]
I think we need it also for other reason.
Let's suppose that we have 200 hypervisor, while we have 2x 100 subscription.
We need to try to attach the first one, but when full we need to switch to the second one.
am i wrong?

btw, i agree that also a "general" registration would be good.
Here another issue raise, we need to have multiple subs to be attached.

to catch both, what about something like the following as structure?
this would complicate the yaml, but allow us to have multiple mandatory subscription.

```
---
settings:
  :user: admin
  :pass: changeme
  :uri: http://localhost
  :org: 1
:subs:
  - 
    hostname: "esxi[0-9]\\.example\\.com"
    registered_by: 85e65e06-a117-4e8e-8aa1-72cb1e00b930
    sub: b9548e4c9fa20b85f264fbaa2470b726
  - 
    hostname: esxi123\.example\.com
    registered_by: ~
    sub: 
      - 
        - 4543828edcf35158c30abc3554c1e36a
        - 5543828edcf35158c30abc3554c1e36b
      - 
        - 6543828edcf35158c30abc3554c1e36c
        - 7543828edcf35158c30abc3554c1e36d
```

Daniele

**Comment by evgeni** [comment]
Hey,

I must admit I think your structure is too complicated. What would happen if the script tries to add `4543828edcf35158c30abc3554c1e36a`, succeeds, then tries `5543828edcf35158c30abc3554c1e36b` and fails? Should `4543828edcf35158c30abc3554c1e36a` be removed again?

I think for 200 hypervisors and 2x100 sub pools, I would suggest to have something like this:

```
:subs:
  - hostname: esxi0[0-9][0-9]\.example\.com
    sub: a9548e4c9fa20b85f264fbaa2470b721
  - hostname: esxi1[0-9][0-9]\.example\.com
    sub: b9548e4c9fa20b85f264fbaa2470b722
```

This is more explicit and removes any guessing the script has to do.

For multiple subs I'd then do:

```
:subs:
  - hostname: esxi0[0-9][0-9]\.example\.com
    sub:
      - a9548e4c9fa20b85f264fbaa2470b721
      - a9548e4c9fa20b85f264fbaa2470b722
  - hostname: esxi1[0-9][0-9]\.example\.com
    sub:
      - b9548e4c9fa20b85f264fbaa2470b722
      - b9548e4c9fa20b85f264fbaa2470b723
```

Now there is also sometimes the need to give the same sub multiple times to a host (e.g. when the sub is only for 2 sockets and we have 4). Not sure how I would like to map that into the YAML.

**Comment by Rocco83** [comment]
Hi,

as by chat, this structure should fit the best.

```
---
settings:
  :user: admin
  :pass: changeme
  :uri: http://localhost
  :org: 1
:subs:
  - 
    hostname: "esxi[0-9]\\.example\\.com"
    registered_by: 85e65e06-a117-4e8e-8aa1-72cb1e00b930
    sub: b9548e4c9fa20b85f264fbaa2470b726
  - 
    hostname: esxi123\.example\.com
    registered_by: ~
    sub: 
      rhel:
        - 4543828edcf35158c30abc3554c1e36a
        - 5543828edcf35158c30abc3554c1e36b
      jboss:
        - 6543828edcf35158c30abc3554c1e36c
        - 7543828edcf35158c30abc3554c1e36d
      satellite:
        - 7543828edcf35158c30abc3554c1e36e
```

PR will follow to fullfill this structure.

**Comment by CalvinHartwell** [comment]
@Rocco83 did you finish this or are you still working on it? I'm about to modify the script for this functionality now. 

I think the structure here is still too simple, I've been to a customer who had very random subs, instead of 2 x 100 RHEL subs they had something like 5-10 different subs (with different IDs) for varying amounts of RHEL machines, all with different expiry dates. Distributing those subs over 100 hypervisors makes for an interesting yaml file. 

Perhaps we need some logic to check if a subscription is already completely utilised? 

Cheers

**Comment by evgeni** [comment]
Hey,

so having a check if there are still enough free subs in the pool before assigning them: :+1:

having the script select a sub from a set of subs... idk, personally I don't like this as I prefer to being explicit, but would not refuse a patch either :)

**Comment by Rocco83** [comment]
Hi Calvin, Evgeni,

this is the purpose of https://github.com/RedHatSatellite/katello-attach-subscription/issues/5 :)
better to move on that issue, imo

**Comment by CalvinHartwell** [comment]
@Rocco83 @evgeni did we fix this?

**Comment by Rocco83** [comment]
Ciao Calvin,

The check in order to skip fully used subscription is fixed in the current master already available.
See cbdaed04d8ab57ad4f0da30e8ef0a914d328b69c

The multiple subs is fixed in my pr, still pending. See #7

What is not fixed is the number of subs to be associated with any subs, nor automatic (candlepin to be patched) or manual (yaml with number of subs fixed for every section) way.
Again, this is #5.
Feel free to patch my master code :-)

Do you have something else still to be implemented?

HTH

**Comment by evgeni** [comment]
closed via 06a27a8897c60d1fe92e252137a2274ba34da817
